plugins/microchip/switchtec-nvme.c

```diff
@@ -261,37 +287,141 @@ static int pax_get_nvme_pf_functions(struct pax_nvme_device *pax,
 	for (i = 0; i < pax_all.ep_port_all.ep_port_count; i++) {
 		ep_port = &pax_all.ep_port_all.ep_ports[i];
 		if (ep_port->port_hdr.type == SWITCHTEC_GFMS_DB_TYPE_EP) {
-			n = ep_port->ep_ep.ep_hdr.function_number;
+			ep = &ep_port->ep_ep;
+			ret = pax_enum_ep(ep, functions + index,
+					  max_functions - index);
+			index += ret;
+		} else if (ep_port->port_hdr.type ==
+			   SWITCHTEC_GFMS_DB_TYPE_SWITCH) {
+			n = ep_port->port_hdr.ep_count;
+
 			for (j = 0; j < n; j++) {
-				function = &ep_port->ep_ep.functions[j];
-				if (function->sriov_cap_pf == CAP_PF
-				    && function->device_class == NVME_CLASS)
-					memcpy(&functions[index++],
-						function,
-						sizeof(*function));
+				ep = ep_port->ep_switch.switch_eps + j;
+				ret = pax_enum_ep(ep, functions + index,
+						  max_functions - index);
+
+				index += ret;
 			}
 		}
 	}
 
 	return index;
 }
 
+static int pax_build_nvme_pf_list(struct pax_nvme_device *pax,
+				  struct switchtec_gfms_db_ep_port_attached_device_function *functions,
+				  int function_n,
+				  struct list_item *list_items,
+				  char *path)
+{
+	int j;
+	int k;
+	int ret;
+	int idx = 0;
+	char node[300];
+	__u32 ns_list[1024];
+
+	for (j = 0; j < function_n; j++) {
+		pax->pdfid = functions[j].pdfid;
+		ret = switchtec_ep_tunnel_status(pax->dev, pax->pdfid,
+						 &pax->channel_status);
+		if (ret)
+			switchtec_perror("Getting EP tunnel status");
+
+		if (pax->channel_status == SWITCHTEC_EP_TUNNEL_DISABLED)
+			switchtec_ep_tunnel_enable(pax->dev, pax->pdfid);
+
+		ret = nvme_identify_ns_list(0, 0, 1, ns_list);
+		if (!ret) {
+			for (k = 0; k < 1024; k++)
+				if (ns_list[k]) {
+					sprintf(node, "0x%04hxn%d@%s",
+						pax->pdfid, ns_list[k], path);
+					pax->ns_id = ns_list[k];
+					ret = get_nvme_info(0,
+							    &list_items[idx++],
+							    node);
+				}
+		} else if (ret > 0) {
+			fprintf(stderr, "NVMe Status:%s(%x) NSID:%d\n",
+					nvme_status_to_string(ret), ret, 0);
+		} else {
+			perror("id namespace list");
+		}
+
+		if (pax->channel_status == SWITCHTEC_EP_TUNNEL_DISABLED)
+			switchtec_ep_tunnel_disable(pax->dev, pax->pdfid);
```

Comment:
disable?

Reply:
Yes this restores the device EP tunnel setting. Otherwise running `switchtec-nvme` leaves all device EP tunnel enabled.